bitwarden_license/src/Commercial.Core/SecretsManager/Queries/ServiceAccounts/CountNewServiceAccountSlotsRequiredQuery.cs

```diff
@@ -0,0 +1,39 @@
+﻿using Bit.Core.Exceptions;
+using Bit.Core.Repositories;
+using Bit.Core.SecretsManager.Queries.ServiceAccounts.Interfaces;
+using Bit.Core.SecretsManager.Repositories;
+
+namespace Bit.Commercial.Core.SecretsManager.Queries.ServiceAccounts;
+
+public class CountNewServiceAccountSlotsRequiredQuery : ICountNewServiceAccountSlotsRequiredQuery
+{
+    private readonly IOrganizationRepository _organizationRepository;
+    private readonly IServiceAccountRepository _serviceAccountRepository;
+
+    public CountNewServiceAccountSlotsRequiredQuery(
+        IOrganizationRepository organizationRepository,
+        IServiceAccountRepository serviceAccountRepository)
+    {
+        _organizationRepository = organizationRepository;
+        _serviceAccountRepository = serviceAccountRepository;
+    }
+
+    public async Task<int> CountNewServiceAccountSlotsRequiredAsync(Guid organizationId, int serviceAccountsToAdd)
+    {
+        var organization = await _organizationRepository.GetByIdAsync(organizationId);
+        if (organization == null || !organization.UseSecretsManager)
+        {
+            throw new NotFoundException();
+        }
+
+        if (!organization.SmServiceAccounts.HasValue)
+        {
+            return 0;
+        }
+
+        var serviceAccountCount = await _serviceAccountRepository.GetServiceAccountCountByOrganizationIdAsync(organizationId);
+        var availableServiceAccountSlots = Math.Max(0, organization.SmServiceAccounts.Value - serviceAccountCount);
```

Comment:
Would this ever be less than zero?

If anything, if these values get out of step, we probably want to correct that here. e.g.

* If an org has 5 slots but 10 current service accounts, then `availableServiceAccountSlots == -5`
* Then assume we want to add 2 service accounts, then the return value will be `2 - -5 == 7`, which means we add what we need (2) and plus what's already being used (5). But that depends on allowing `availableServiceAccountSlots` to be negative.

Reply:
That makes sense, I've revised the logic, please re-review.

Reply:
lgtm